File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_CLANMEMBERSTATUSCHANGE.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_CLANMEMBERSTATUSCHANGE : Message
    {
        public SID_CLANMEMBERSTATUSCHANGE()
        {
            Id = (byte)MessageIds.SID_CLANMEMBERSTATUSCHANGE;
            Buffer = new byte[6];
        }

        public SID_CLANMEMBERSTATUSCHANGE(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_CLANMEMBERSTATUSCHANGE;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            if (context.Direction != MessageDirection.ServerToClient)
            {
                Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
                throw new GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be sent from server to client");
            }

            /**
             * (STRING) Username
             *  (UINT8) Rank
             *  (UINT8) Online status
             * (STRING) Location
             */

            byte[] username = (byte[])context.Arguments["username"];
            byte rank = (byte)context.Arguments["rank"];
            byte status = (byte)context.Arguments["status"];
            byte[] location = (byte[])context.Arguments["location"];

            Buffer = new byte[4 + username.Length + location.Length];

            using var m = new MemoryStream(Buffer);
            using var w = new BinaryWriter(m);

            w.WriteByteString(username);
            w.Write(rank);
            w.Write(status);
            w.WriteByteString(location);

            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)");
            context.Client.Send(ToByteArray(context.Client.ProtocolType));
            return true;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_CLANMEMBERSTATUSCHANGE.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_CLANMEMBERSTATUSCHANGE
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_CLANMEMBERSTATUSCHANGE)
            Buffer = New Byte(5) {}
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_CLANMEMBERSTATUSCHANGE)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            If context.Direction <> MessageDirection.ServerToClient Then
                Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} must be sent from server to client")
            End If

            Dim username As Byte() = CType(context.Arguments("username"), Byte())
            Dim rank As Byte = CByte(context.Arguments("rank"))
            Dim status As Byte = CByte(context.Arguments("status"))
            Dim location As Byte() = CType(context.Arguments("location"), Byte())
            Buffer = New Byte(4 + username.Length + location.Length - 1) {}

            Using m = New MemoryStream(Buffer)
                Using w = New BinaryWriter(m)
                    w.WriteByteString(username)
                    w.Write(rank)
                    w.Write(status)
                    w.WriteByteString(location)
                End Using
            End Using

            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
            context.Client.Send(ToByteArray(context.Client.ProtocolType))
            Return True
        End Function
    End Class
End Namespace
